File: algorithm/trapping-rain-water.cpp
42. Trapping Rain Water
Difficulty: Hard

Given n non-negative integers representing an elevation map where the width of each bar is 1, 
compute how much water it is able to trap after raining.

For example, 
Given [0,1,0,2,1,0,1,3,2,1,2,1], return 6.

// Time Complexity: O(n)
// Space Complexity: O(1)

class Solution {
public:
    int trap(int A[], int n) {
        const int n = A.size();
        stack<pair<int, int>> s;
        int water = 0;
        for (int i = 0; i < n; ++i) {
            int height = 0;
            while (!s.empty()) {
                int bar = s.top().first;
                int pos = s.top().second;
                water += (min(bar, A[i]) - height) * (i - pos - 1);
                height = bar;
                if (A[i] < bar) {
                    break;
                } else {
                    s.pop();
                }
            }
            s.push(make_pair(A[i], i));
        }
        return water;
    }
};

File: algorithm/oc/trapping-rain-water.m
// 42. Trapping Rain Water
// Difficulty: Hard

// Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.

// For example, 
// Given [0,1,0,2,1,0,1,3,2,1,2,1], return 6.

#import <Foundation/Foundation.h>

int trap(int* height) {
  int left = 0, right = height.length - 1, level = 0, res = 0;
  while (left < right) {
      int lower = height[(height[left] < height[right]) ? left++ : right--];
      level = MAX(level, lower);
      res += level - lower;
  }
  return res;
}